mailpoet/lib/Validator/Schema/ObjectSchema.php

```diff
@@ -0,0 +1,55 @@
+<?php declare(strict_types = 1);
+
+namespace MailPoet\Validator\Schema;
+
+use MailPoet\Validator\Schema;
+
+// See: https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#objects
+class ObjectSchema extends Schema {
+  protected $schema = [
+    'type' => 'object',
+  ];
+
+  /** @param array<string, Schema> $properties */
+  public function properties(array $properties): self {
+    return $this->updateSchemaProperty('properties', array_map(
+      function (Schema $property) {
+        return $property->toArray();
+      },
+      $properties
+    ));
+  }
+
+  public function additionalProperties(Schema $schema): self {
+    return $this->updateSchemaProperty('additionalProperties', $schema->toArray());
+  }
+
+  public function enableAdditionalProperties(): self {
+    if (($this->schema['additionalProperties'] ?? []) === false) {
+      return $this->unsetSchemaProperty('additionalProperties');
+    }
+    return clone $this;
+  }
+
+  public function disableAdditionalProperties(): self {
+    return $this->updateSchemaProperty('additionalProperties', false);
+  }
+
+  /** @param array<string, Schema> $properties */
+  public function patternProperties(array $properties): self {
```

Comment:
That's a good idea. Strictly following JSON schema specs we should validate for ECMA 262 regex specs but I'm not sure if there's an easy way to do that (e.g. a regex that validates ECMA 262 regex), can't find any with a quick search.

But maybe... since we're in PHP and WP uses PHP regexes internally anyway, we could simply try to match it against an empty string and catch any errors. I think that would be a good addition 👍

Reply:
Added a new commit to address this for both string patterns and pattern properties.